### PR TITLE
Remove inheritance from base_client_integration_test to engine_builder

### PR DESCRIPTION
**Pull request by caschoener**
Currently the unit test fixture inherits from the engine builder. This is mostly just a readability win and also allows setting up multiple builders within one test.

### CODE REVIEW

File: mobile/library/cc/engine_builder.h [hunk 1]
```diff
@@ -66,8 +66,6 @@ class EngineBuilder {
   std::string generateConfigStr() const;
 
   EngineSharedPtr build();
-
-protected:
```
Comment (abeyad) [inline mobile/library/cc/engine_builder.h:70]: I'm weary of making these two functions available to any user of the EngineBuilder class.  Can we keep it protected (or private) and make BaseClientIntegrationTest a friend class of EngineBuilder? https://en.cppreference.com/w/cpp/language/friend

Friendship is not inherited in C++, but we can get around that by adding functions to BaseClientIntegrationTest that call into the EngineBuilder's protected functions, and have ClientIntegrationTest and XdsIntegrationTest call into those.

Comment (caschoener) [reply mobile/library/cc/engine_builder.h:70]: Learn something new about c++ every day!  Sounds good.

File: mobile/library/cc/engine_builder.h [hunk 2]
```diff
@@ -16,6 +16,8 @@ namespace Envoy {
 namespace Platform {
 
 class EngineBuilder {
+  friend Envoy::BaseClientIntegrationTest;
```
Comment (abeyad) [inline mobile/library/cc/engine_builder.h:19]: can we move the friend declaration to the protected section?

**Review by abeyad — APPROVED** [review]

**Comment by abeyad** [comment]
this has merge conflicts with main that need to be resolved

**Comment by abeyad** [comment]
Added @jpsim for cross-company review

**Review by jpsim — APPROVED** [review]

**Pull request merged**